### PR TITLE
Features/#300 unused log section

### PR DESCRIPTION
**Pull request by dakochik**
Issue #300 is resolved. Now inspection collects all references from SL injections and checks if there any reference to log section, instead of using regular expressions.

It is using `SnakemakeInspectionVisitor` instead of ` SmkSLInspectionVisitor` because in the second case it is less convenient to find and highlight exact unused `log` sections.

### CODE REVIEW

File: src/main/kotlin/com/jetbrains/snakecharm/lang/psi/SmkSLSectionReferencesCollector.kt [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        subscriptions?.forEach { st ->
+            val child = (st.node.firstChildNode.psi as? SmkSLReferenceExpressionImpl)
+            if (child != null) {
+                checkSLReference(child)
+            }
+        }
+    }
+
+    private fun checkSLReference(ref: SmkSLReferenceExpressionImpl) {
+        if (ref.isWildcard()) {
```
Comment (iromeo) [inline src/main/kotlin/com/jetbrains/snakecharm/lang/psi/SmkSLSectionReferencesCollector.kt:67]: AFAIU you ignore '{log}' where log is wildcard, so isn't considered as log sections usages (that is reasonable), but this case seems not covered by tests

File: src/main/kotlin/com/jetbrains/snakecharm/lang/psi/SmkSLSectionReferencesCollector.kt [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+import com.jetbrains.python.psi.PyRecursiveElementVisitor
+import com.jetbrains.python.psi.PyStringLiteralExpression
+import com.jetbrains.snakecharm.lang.psi.elementTypes.SmkElementTypes
+import com.jetbrains.snakecharm.stringLanguage.lang.psi.SmkSLFile
+import com.jetbrains.snakecharm.stringLanguage.lang.psi.SmkSLReferenceExpressionImpl
+import com.jetbrains.snakecharm.stringLanguage.lang.psi.SmkSLSubscriptionExpressionImpl
+
+class SmkSLSectionReferencesCollector(
+    private val searchingType: String,
+    private val isRuleOrCheckpointAppropriate: (SmkRuleOrCheckpoint) -> Boolean = { true }
```
Comment (iromeo) [inline src/main/kotlin/com/jetbrains/snakecharm/lang/psi/SmkSLSectionReferencesCollector.kt:20]: For `use` syntax exection section will be missing, so we will likely got a FP on things like
```shell
rule foo:
   shell: "echo {log}"  # inspection warning here could be ignored, it is already reported, no error on runime, undefined sections are empty listrs

use rule foo as boo:
   log: "aa.log"  # <- likely FP here
```
so please add an issue for `use` support to cover uch case in this inspection.

File: src/main/kotlin/com/jetbrains/snakecharm/inspections/SmkUnusedLogSectionInspection.kt [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        }
+
+        private fun visitSMKRuleLike(rule: SmkRuleLike<SmkArgsSection>) {
+            val logSection = rule.getSections().firstOrNull {
+                it.sectionKeyword == SnakemakeNames.SECTION_LOG
+            } ?: return
+            val sectionsByRule = session.putUserDataIfAbsent(KEY, hashSetOf())
+
+            if (rule in sectionsByRule) {
+                return
```
Comment (iromeo) [inline src/main/kotlin/com/jetbrains/snakecharm/inspections/SmkUnusedLogSectionInspection.kt:38]: Not sure that is actually needed. What is used for? Is it caching that allows next time not to check the rule? What if rule has changed? Inspections will be re-run automatically, but cache could be out-of-data. Not 100% sure, just my guess. Maybe I didn't get the idea.

File: src/main/kotlin/com/jetbrains/snakecharm/inspections/SmkUnusedLogSectionInspection.kt [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+            }
+
+            val collector = SmkSLSectionReferencesCollector(SnakemakeNames.SECTION_LOG) {
+                !sectionsByRule.contains(rule)
```
Comment (iromeo) [inline src/main/kotlin/com/jetbrains/snakecharm/inspections/SmkUnusedLogSectionInspection.kt:42]: Let's rewrite 3 statements as one, like:
```
SmkSLSectionReferencesCollector() {
..
}.also {
  rule.containing.file.accept(it)
}.getSections().foreach() { ...}
```

Comment (iromeo) [inline src/main/kotlin/com/jetbrains/snakecharm/inspections/SmkUnusedLogSectionInspection.kt:42]: Why `!sectionsByRule.contains(rule)` doesn't use lambda argument `it` here? is it a bug? If no, why do we need lambda, we could replace this with just boolean value and pass into constructor.

**Review by iromeo — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/jetbrains/snakecharm/inspections/SmkUnusedLogSectionInspection.kt (hunk 3 above)
Comment (dakochik) [reply src/main/kotlin/com/jetbrains/snakecharm/inspections/SmkUnusedLogSectionInspection.kt:38]: As I understand, `sectionsByRule` will be overwritten before each inspection call (in debug mode, `session` saves rules we need while it checks file rules and then, befor next file inspection, it forgots all rules). During one inspection session code can't be modified, so at this point everything should be ok. 
Every time inspection visits rule statement, it scans all available rules, so we use such check in order not to call scanning (via `SmkSLSectionReferencesCollector`) for rules about which we know, that they were referenced during current inspection session

File: src/main/kotlin/com/jetbrains/snakecharm/lang/psi/SmkSLSectionReferencesCollector.kt [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+            }
+        }
+    }
+
+    private fun checkSLReference(ref: SmkSLReferenceExpressionImpl) {
+        if (ref.isWildcard()) {
+            return
+        }
+        var element = ref.reference.resolve()
+        if (element.elementType == PyElementTypes.KEYWORD_ARGUMENT_EXPRESSION) {
```
Comment (iromeo) [inline src/main/kotlin/com/jetbrains/snakecharm/lang/psi/SmkSLSectionReferencesCollector.kt:71]: Just if `(element is PyKeywordArgument)` seems we don't need low level access to AST node here. ?

File: src/main/kotlin/com/jetbrains/snakecharm/lang/psi/SmkSLSectionReferencesCollector.kt [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+            element = (element as PyKeywordArgument).parentOfType<SmkRuleOrCheckpointArgsSection>()
+        }
+        if (
+            (element.elementType == SmkElementTypes.RULE_OR_CHECKPOINT_ARGS_SECTION_STATEMENT
```
Comment (iromeo) [inline src/main/kotlin/com/jetbrains/snakecharm/lang/psi/SmkSLSectionReferencesCollector.kt:75]: Same question as before, why do we need AST tree node type access. Seems PSI is enough here.

File: src/main/kotlin/com/jetbrains/snakecharm/lang/psi/SmkSLSectionReferencesCollector.kt [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+                ?.map { it.first }
+                ?.filterIsInstance<SmkSLFile>()
+                ?: return
+
+        injectedFiles.forEach(::collectResolvedReferences)
+    }
+
+    private fun collectResolvedReferences(file: SmkSLFile) {
+        val references = PsiTreeUtil.getChildrenOfType(file, SmkSLReferenceExpressionImpl::class.java)
+        references?.forEach { st ->
```
Comment (iromeo) [inline src/main/kotlin/com/jetbrains/snakecharm/lang/psi/SmkSLSectionReferencesCollector.kt:54]: Please avoid using `FooImpl`, used corresponding interface `Foo`

**Review by iromeo — CHANGES_REQUESTED** [review]

**Comment by dakochik** [comment]
fixed: now we check if current rule (not any other) refer to `log` section, new quick fixes, additional tests, code refactoring

**Review by iromeo — APPROVED** [review]
In general is good, some things could be improved. No action required. PR is accepted.

**Pull request merged**